saleor/graphql/api.py

```diff
@@ -168,3 +181,35 @@ def serialize_webhook_event(value):
     directives=graphql.specified_directives
     + [GraphQLDocDirective, GraphQLWebhookEventsInfoDirective],
 )
+
+
+def _fail(errors, **_kwargs) -> ExecutionResult:
+    return ExecutionResult(errors=errors, invalid=True)
+
+
+class SaleorGraphQLBackend(GraphQLCoreBackend):
+    def document_from_string(
+        self,
+        schema: GraphQLSchema,
+        document_string: str,  # type: ignore[override]
+    ) -> GraphQLDocument:
+        # validate eagerly so we can cache the result
+        document_ast = parse(document_string)
+        validation_errors = validate(schema, document_ast)
+        if validation_errors:
+            return GraphQLDocument(
+                schema=schema,
+                document_string=document_string,
+                document_ast=document_ast,
+                execute=partial(_fail, validation_errors),
```

Comment:
So the purpose here is to have a different `execute` function that returns early (when we know there are validation issues), without going into all logic that the regular `execute` does?

Reply:
Yes, the previous implementation, `validate_and_execute` from `graphql-core`, did precisely that: it pretended to be `execute`, but internally first ran `validate` and if that failed, immediately returned a result without calling `execute`.

Reply:
Ideally, we would not even bother with caching incorrect queries (as they are unlikely to be repeated), but that would require writing an even more elaborate backend implementation. Anything uncommon will be pushed out by more popular queries so it should be fine as it is.